src/charm.py

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env python3
+
+import os
+import subprocess
+import socket
+import re
+import pwd
+import sys
+sys.path.append('lib')  # noqa
+
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+    WaitingStatus,
+    ModelError,
+)
+from cluster import CockroachDBCluster
+from interface_proxy_listen_tcp import ProxyListenTcpInterfaceProvides
+
+from jinja2 import Environment, FileSystemLoader
+from datetime import timedelta
+from time import sleep
+
+
+class CockroachStartedEvent(EventBase):
+    pass
+
+
+class ClusterInitializedEvent(EventBase):
+    def __init__(self, handle, cluster_id):
+        super().__init__(handle)
+        self.cluster_id = cluster_id
+
+    def snapshot(self):
+        return self.cluster_id
+
+    def restore(self, cluster_id):
+        self.cluster_id = cluster_id
+
+
+class CockroachDBCharmEvents(CharmEvents):
+    cockroachdb_started = EventSource(CockroachStartedEvent)
+    cluster_initialized = EventSource(ClusterInitializedEvent)
+
+
+class CockroachDBCharm(CharmBase):
+    on = CockroachDBCharmEvents()
+    state = StoredState()
+
+    COCKROACHDB_SERVICE = 'cockroachdb.service'
+    SYSTEMD_SERVICE_FILE = f'/etc/systemd/system/{COCKROACHDB_SERVICE}'
+    WORKING_DIRECTORY = '/var/lib/cockroach'
+    COCKROACH_INSTALL_DIR = '/usr/local/bin'
+    COCKROACH_BINARY_PATH = f'{COCKROACH_INSTALL_DIR}/cockroach'
+    COCKROACH_USERNAME = 'cockroach'
+    PSQL_PORT = 26257
+    HTTP_PORT = 8080
+
+    MAX_RETRIES = 10
+    RETRY_TIMEOUT = timedelta(milliseconds=125)
+
+    def __init__(self, framework, key):
+        super().__init__(framework, key)
+
+        self.state.set_default(is_started=False)
+
+        for event in (self.on.install,
+                      self.on.start,
+                      # self.on.upgrade_charm,
+                      self.on.config_changed,
+                      self.on.cluster_relation_changed,
+                      self.on.cockroachdb_started,
+                      self.on.proxy_listen_tcp_relation_joined):
+            self.framework.observe(event, self)
```

Comment:
The fact everything is handled by this type alone makes me feel like there's a lack of encapsulation, and consequent lack of code reusability. Not sure if that's the case at this point yet.

proxy-listen-tcp is also a weird interface. "proxying" is in general a high-level behavior for an underlying protocol. I'd expect the interface to be about the protocol, not about how it's being used by the server or the client. But maybe I don't understand something here.

Reply:
I agree, the discussion here https://github.com/canonical/haproxy-operator/pull/1#issuecomment-607500253 indicates to me that I could try and come up with abstractions that would cover tcp load-balancing taking common properties of, say, haproxy, LVS and nginx into account. That should allow me to factor out certain logic unrelated to this application to different types.

You are right, naming needs to be improved, commented here: https://github.com/canonical/cockroachdb-operator/pull/1#discussion_r402943185

Reply:
I moved a lot of functionality to a separate type to support unit testing.

https://github.com/canonical/cockroachdb-operator/pull/1/commits/e6741a9a01432f19d445914a77ff3a65050f0bcd

I also removed the support for `proxy-listen-tcp` interface for now so we can focus on getting the initial setup + clustering reviewed separately from the tcp-load-balancer interface PR (after it is done I can add it back given it is a small delta).